x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```
@@ -408,43 +490,95 @@ public void authenticate(final AuthenticationToken authenticationToken, final Ac
+                    jwtHeader.getAlgorithm().getName()
                 );
-                claimsSet = jwt.getJWTClaimsSet();
-                LOGGER.trace("Realm [{}] JWT validation succeeded for token=[{}].", super.name(), tokenPrincipal);
+                JwtValidateUtil.validateType(jwt);
+                JwtValidateUtil.validateIssuer(jwt, allowedIssuer);
+                JwtValidateUtil.validateAudiences(jwt, allowedAudiences);
+                JwtValidateUtil.validateSignatureAlgorithm(jwt, isJwtSignatureAlgHmac ? this.allowedJwksAlgsHmac : this.allowedJwksAlgsPkc);
+                JwtValidateUtil.validateAuthTime(jwt, now, this.allowedClockSkew.seconds());
+                JwtValidateUtil.validateIssuedAtTime(jwt, now, this.allowedClockSkew.seconds());
+                JwtValidateUtil.validateNotBeforeTime(jwt, now, this.allowedClockSkew.seconds());
+                JwtValidateUtil.validateExpiredTime(jwt, now, this.allowedClockSkew.seconds());
+
+                // Validate signature last, so expensive JWK reloads only happen after all other checks passed.
+                try {
+                    if (this.contentAndFilteredJwksAlgsHmac.isEmpty() && this.contentAndFilteredJwksAlgsPkc.isEmpty()) {
+                        LOGGER.debug("Realm [" + super.name() + "] has no JWKs and Algs to verify JWT for token=[" + tokenPrincipal + "].");
+                        if (this.reloadJwksHelper(listener, tokenPrincipal, isJwtSignatureAlgHmac) == false) {
+                            return; // Try reload before. Reload failed, or succeeded but no changes. Helper handles listener and logs.
+                        }
+                        this.verifyNonEmptyHmacAndPkcJwksAlgs(); // throws SettingsException if reload causes filtered PKC+HMAC JWKs and
+                                                                 // Algs to drop to 0
+                    }
```

I don't think we need proactively reload here. IIUC, empty keys/algos will just lead to signature validation failure which we catuch later on and does the reload anyway. Removing it makes the flow simpler.

---

Removing reload before validateSignature makes sense, because validateSignature will fail fast for empty PKC JWKs Really good suggestion! I like this a lot.